internal/provider/folder_data_source.go

```diff
@@ -0,0 +1,181 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/cockroachdb/cockroach-cloud-sdk-go/pkg/client"
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ datasource.DataSource              = &folderDataSource{}
+	_ datasource.DataSourceWithConfigure = &folderDataSource{}
+)
+
+type folderDataSource struct {
+	provider *provider
+}
+
+func (d *folderDataSource) Schema(
+	_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse,
+) {
+	resp.Schema = schema.Schema{
+		Description: "A CockroachDB Cloud folder. Folders can contain clusters or other folders.  They can be used to group resources together for the purposes of access control, organization or fine grained invoicing.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				MarkdownDescription: "The id the folder.",
+				Optional:            true,
+			},
+			"path": schema.StringAttribute{
+				MarkdownDescription: "An absolute path to the folder. Trailing slashes are optional. (i.e. /folder1/folder2)",
+				Optional:            true,
+			},
+			"name": schema.StringAttribute{
+				MarkdownDescription: "Name of the cluster.",
+				Computed:            true,
+			},
+			"parent_id": schema.StringAttribute{
+				Computed:            true,
+				MarkdownDescription: "The ID of the folders's parent folder. 'root' is used for a folder at the root level.",
+			},
+		},
+	}
+}
+
+func (d *folderDataSource) Metadata(
+	_ context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse,
+) {
+	resp.TypeName = req.ProviderTypeName + "_folder"
+}
+
+func (d *folderDataSource) Configure(
+	_ context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse,
+) {
+	if req.ProviderData == nil {
+		return
+	}
+	var ok bool
+	if d.provider, ok = req.ProviderData.(*provider); !ok {
+		resp.Diagnostics.AddError("Internal provider error",
+			fmt.Sprintf("Error in Configure: expected %T but got %T", provider{}, req.ProviderData))
+	}
+}
+
+func (d *folderDataSource) Read(
+	ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse,
+) {
+	if d.provider == nil || !d.provider.configured {
+		addConfigureProviderErr(&resp.Diagnostics)
+		return
+	}
+
+	var folderDataSource FolderDataSourceModel
+	diags := req.Config.Get(ctx, &folderDataSource)
+
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		resp.Diagnostics.AddWarning("Error loading the folder", "")
+		return
+	}
+
+	if folderDataSource.ID.IsNull() && folderDataSource.Path.IsNull() {
```

Comment:
Is there a way we can fail earlier here? Maybe on `Configure`?

Reply:
There is a built in validation framework.  I'll look into using that.

Reply:
I've incorporated the validation framework.  On one hand its nice to use something built in but on the other the error messages can be pretty bad.   For parent_id the field either needs to be a uuid or match the string "root".  But the framework gives 2 distinct errors when it doesn't match:


```
➜  cockroach_folder git:(CC-27470-support-new-terraform-data-source-for-folders) ✗ terraform plan
╷
│ Error: Invalid Attribute Value Match
│
│   with cockroach_folder.example_folder_child,
│   on main.tf line 59, in resource "cockroach_folder" "example_folder_child":
│   59:   parent_id = "d01989b0-b005-493d-a0ae-585e7fc93b7"
│
│ Attribute parent_id value must be one of: ["root"], got: "d01989b0-b005-493d-a0ae-585e7fc93b7"
╵
╷
│ Error: Invalid Attribute Value Match
│
│   with cockroach_folder.example_folder_child,
│   on main.tf line 59, in resource "cockroach_folder" "example_folder_child":
│   59:   parent_id = "d01989b0-b005-493d-a0ae-585e7fc93b7"
│
│ Attribute parent_id must match UUID format, got: d01989b0-b005-493d-a0ae-585e7fc93b7
╵
```

I may look a bit more for a better option.

Reply:
I went ahead and added a custom validator for the parent_id.  It was pretty straightforward and improves the experience.  https://github.com/cockroachdb/terraform-provider-cockroach/pull/190/files#diff-c7ceb2bc425e0c0b8e8ad8826979ce04e42f1151285d04332a9fce92ca30f7b3